metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/common/shell/DefaultStellarShellExecutor.java

```diff
@@ -0,0 +1,398 @@
+    this.functionResolver = functionResolver;
+  }
+
+  public DefaultStellarShellExecutor(
+          Properties properties,
+          Optional<String> zookeeperUrl) throws Exception {
+
+    this(StellarFunctions.FUNCTION_RESOLVER(), properties, zookeeperUrl);
+  }
+
+  @Override
+  public void init() {
+    StellarFunctions.initialize(this.context);
+    this.commandRegistry = registerSpecialCommands();
+
+    // TODO this wont really work as functions are probably not defined yet
+    // but the auto-complete reaches out to the function resolver separately
+    for(StellarFunctionInfo fn : functionResolver.getFunctionInfo()) {
+      notifyFunctionListeners(fn);
+    }
+  }
+
+  /**
+   * Add a listener that will be notified when a function is defined.
+   * @param listener The listener to notify.
+   */
+  @Override
+  public void addFunctionListener(FunctionDefinedListener listener) {
+    this.functionListeners.add(listener);
+  }
+
+  /**
+   * Notify function listeners that a function has been defined.
+   * @param functionInfo The function that was defined.
+   */
+  private void notifyFunctionListeners(StellarFunctionInfo functionInfo) {
+    for(FunctionDefinedListener listener : functionListeners) {
+      listener.whenFunctionDefined(functionInfo);
+    }
+  }
+
+  /**
+   * Add a listener that will be notified when a variable is defined.
+   * @param listener The listener to notify.
+   */
+  @Override
+  public void addVariableListener(VariableDefinedListener listener) {
+    this.variableListeners.add(listener);
+  }
+
+  /**
+   * Notify variable listeners that a variable has been (re)defined.
+   * @param variableName The variable name.
+   * @param result The variable result.
+   */
+  private void notifyVariableListeners(String variableName, VariableResult result) {
+    for(VariableDefinedListener listener : variableListeners) {
+      listener.whenVariableDefined(variableName, result);
+    }
+  }
+
+  /**
+   * Add a listener that will be notified when a magic command is defined.
+   * @param listener The listener to notify.
+   */
+  @Override
+  public void addSpecialListener(SpecialDefinedListener listener) {
+    this.specialListeners.add(listener);
+  }
+
+  /**
+   * Notify listeners that a magic command has been defined.
+   * @param specialCommand The magic command.
+   */
+  private void notifySpecialListeners(SpecialCommand specialCommand) {
+    for(SpecialDefinedListener listener : specialListeners) {
+      listener.whenSpecialDefined(specialCommand);
+    }
+  }
+
+  @Override
+  public StellarShellResult execute(String expression) {
+    expression = expression.trim();
+
+    // if whitespace, there is nothing much to do
+    if(StringUtils.isBlank(expression)) {
+      return noop();
+    }
+
+    // is this a special command?
+    for(SpecialCommand command : commandRegistry) {
+      if(command.getMatcher().apply(expression)) {
+        return command.execute(expression, this);
+      }
```

Comment:
Anything that is not "core" Stellar is defined separately as a `SpecialCommand`.